### ISSUE TITLE
Missing error for attempted higher order instance

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#1322]
```
Consider the following snippet:

data _==_ {A : Set} (a : A) : A -> Set where
  refl : a == a

data ⊥ : Set where

data ℕ : Set where
  zero : ℕ
  suc : ℕ -> ℕ

f : (n : ℕ) ⦃ p : n == zero -> ⊥ ⦄ -> ℕ
f n = n

g : (n : ℕ) ⦃ q : n == zero -> ⊥ ⦄ -> ℕ
g n ⦃ q ⦄ = f n

Agda head complains that
"No variable of type ⊥ was found in scope.
when checking that n are valid arguments to a function of type
(n₁ : ℕ) {{_ : n₁ == zero → ⊥}} → ℕ"
that is, it is optimistically trying to solve the instance p : n == zero -> ⊥
in the argument to f by a term of the form \ _ -> b for an instance b : ⊥, 
even
though the instance q : n == zero -> ⊥ is already in the context.

I guess the correct thing to do for higher-order instances of type (x : A) -> B 
is to consider all local variables of the right type (x : A) -> B as instances, 
as well as terms of the form \ x . b, where b is an instance of the output type 
B, but I don't know how expensive this would be, performance-wise?

It might be foolish to search for higher-order instances, but I found this in 
the wild in the HoTT-Agda codebase as a way of propagating the constraint 
automatically.
(This also means that it used to work with the old instance search machinery.)
```

Original issue reported on code.google.com by `fredrik....@gmail.com` on 21 Oct 2014 at 2:56

**Comment by UlfNorell** [comment]
Explicit arguments are now solved by instance search (#1540), so the example with top-level q works.

**Comment by UlfNorell** [comment]
The missing check for higher-order instances causes more serious problems. If you try to fill in the instance using `it` you get an internal error:

```
it : ∀ {a} {A : Set a} ⦃ x : A ⦄ → A
it ⦃ x ⦄ = x

h : (n : ℕ) ⦃ q : n == zero → ⊥ ⦄ → ℕ
h n ⦃ q ⦄ = f n ⦃ it ⦄
-- An internal error has occurred. Please report this as a bug.
-- Location of the error: src/full/Agda/TypeChecking/InstanceArguments.hs:277
```

At the moment I don't understand why you get an internal error in this case and a failed instance search in the original example.

**Comment by UlfNorell** [comment]
The fix to #2034 also fixes the internal error above, giving you the same failed instance search as the OP.

When looking for a goal of function type, instance search sticks the argument in the context and searches for an instance of the target type. In the example above this fails because the `n == zero` argument that's added to the context when search for an instance of `n == zero → ⊥` isn't eligible for instance search. Note that the following variant type checks:

``` agda
data _==_ {A : Set} (a : A) : A → Set where
  refl : a == a

data ⊥ : Set where

data ℕ : Set where
  zero : ℕ
  suc : ℕ → ℕ

it : ∀ {a} {A : Set a} ⦃ x : A ⦄ → A
it ⦃ x ⦄ = x

f : (n : ℕ) ⦃ p : ⦃ _ : n == zero ⦄ → ⊥ ⦄ → ℕ
f n = n

g : (n : ℕ) ⦃ q : ⦃ _ : n == zero ⦄ → ⊥ ⦄ → ℕ
g n ⦃ q ⦄ = f n

h : (n : ℕ) ⦃ q : ⦃ _ : n == zero ⦄ → ⊥ ⦄ → ℕ
h n ⦃ q ⦄ = f n ⦃ it ⦄
```

I'm tempted to close this as _working as intended_, but I'll open an issue for instance argument failure diagnosis, because that could certainly be improved.

**Comment by asr** [comment]
> The fix to #2034 also fixes the internal error above

Since #2034 has 2.5.1.1 milestone, I assigned the same milestone to this issue.

**Comment by Wikigoagec** [comment]
/KPS_9566 580-coal- case Clock I told blessiness that go years.
